### DISCUSSION TITLE
custom styling of WFS

### DISCUSSION THREAD

**Discussion by uwer** [TerriaJS/terriajs#7039] [Q&A]
Hi guys
is it possible to customise styling of point layer via WFS beyond the simple maki icons? I am in particular thinking of clustering points and adjusting size based on attribute - client side.

**Comment by nf-s** [comment]
Hi @uwer 

I am going to transfer this issue to the TerriaJS repository and convert it into a discussion post.

We currently do not have the ability to cluster points

- Please see https://github.com/TerriaJS/terriajs/issues/695

We do have the ability to adjust size based on attributes client-side. You can set these attributes in the "Edit Style" workflow

- Load WFS layer
- Open options menu and select "Edit Style"
- <img width="427" alt="image" src="https://github.com/TerriaJS/TerriaMap/assets/6187649/b47a5c1e-1f8e-4b4a-a30a-f0b0c25ddad6">
- Select "Point/Marker style" in the "Symbology" drop-down
- <img width="351" alt="image" src="https://github.com/TerriaJS/TerriaMap/assets/6187649/ea8f1131-583a-4355-ba3b-4172a09ff105">
- Then you can specify marker height/width - this can also be based on an attribute - to do this select a "Variable"
- You can also specify point size for numeric attributes - to do this select "Point size" in the "Symbology" drop-down
- <img width="358" alt="image" src="https://github.com/TerriaJS/TerriaMap/assets/6187649/d6f8c283-1d8e-424a-88d4-35d4239ccff1">

**Reply by uwer** [reply]
Thanks Nick,
(especially for the quick reply)

The interactive styling is not an option here, i will need to predefine the styling to achieve the relevant effect. Is it possible to provide the necessary attributes under the catalogue item? 

The previous request is quite old, Leaflet has extension supporting clustering but i could see that that may not be an option if the intention is to drop Leaflet in favour of Cesium2D.
Cesium appears to support clustering points, would forcingCesium features help?

Cheers
Uwe

**Reply by nf-s** [reply]
Hi @uwer 

You can use the "Edit Style" workflow to generate Terria catalog item JSON

1. Click the `Show advanced options` menu at the top
    ![image](https://github.com/TerriaJS/terriajs/assets/1430/443e3dbf-625c-401f-a06b-366159f25edd)

2. ...and then click again and choose copy-to-clipboard option. This will copy the style definition to clipboard which you can then use in your catalog file.

    ![image](https://github.com/TerriaJS/terriajs/assets/1430/e83bdcef-3683-4eed-a2a1-4aa6298540c6)

As for adding clustering support, it would be unlikely we would use a Leaflet plugin as Terria's vector rendering is tightly coupled with Cesium. The current approach is that we basically re-implement parts of cesium `DataSource` in Leaflet.

We are currently looking at dropping Leaflet in favour of Cesium 2d - which would make adding clustering support relatively easy - I estimate it would take less than a week of dev time.

**Reply by uwer** [reply]
Thanks Nick, that is good news and i think Giovanni might appreciate that too. 
I will pick this up again with you when upcoming project details are discussed.
Cheers
Uwe

**Comment by glughi** [comment]
Hello, you can use a combination of `style` and `perPropertyStyles` to manage the size (and other properties) of your items based on an attribute, see [WebFeatureServiceCatalogItem](https://docs.terria.io/guide/connecting-to-data/catalog-type-details/wfs/).

As for clustering, I needed it too, but it is not present by default in Terria. I activated Cesium’s with very few changes to Terria’s code.

I have also replaced Leaflet with Cesium 2D to have the same behaviour and clustering in 3D and 2D.

**Reply by glughi** [reply]
```
if (this.clusterize) {
    const pinBuilder = new PinBuilder();
    dataSource.clustering.enabled = true;
    dataSource.clustering.pixelRange = 35;
    dataSource.clustering.minimumClusterSize = 5;
    dataSource.clustering.clusterEvent.addEventListener(function (
      entities,
      cluster
    ) {
      cluster.label.show = false;
      cluster.billboard.verticalOrigin = VerticalOrigin.BOTTOM;
      cluster.billboard.image = pinBuilder
        .fromText(entities.length.toLocaleString(), Color.GRAY, 60)
        .toDataURL();
      cluster.billboard.show = true;
    });
  }
```

This is a very simple example of using clustering.
